### PR TITLE
Add package py-devito and it's dependencies

### PR DESCRIPTION
**Pull request by stephenmsachs**
Dependencies include:
- py-cgen
- py-codepy
- py-contexttimer
- py-nbval
- py-pyrevolve

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-cgen/package.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    """
+
+    homepage = "https://documen.tician.de/cgen/"
+    pypi     = "cgen/cgen-2020.1.tar.gz"
+
+    version('2020.1', sha256='4ec99d0c832d9f95f5e51dd18a629ad50df0b5464ce557ef42c6e0cd9478bfcf')
+
+    depends_on('python@:3.5,3.7:', type=('build', 'run'))
+    depends_on('py-pytools@2015.1.2:', type='build')
+    depends_on('py-numpy@1.6:', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cgen/package.py]: These should be type=build/run

File: var/spack/repos/builtin/packages/py-cgen/package.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    """cgen offers a simple abstract syntax tree for C and related languages
+    (C++/CUDA/OpenCL) to allow structured code generation from Python.
+    """
+
+    homepage = "https://documen.tician.de/cgen/"
+    pypi     = "cgen/cgen-2020.1.tar.gz"
+
+    version('2020.1', sha256='4ec99d0c832d9f95f5e51dd18a629ad50df0b5464ce557ef42c6e0cd9478bfcf')
+
+    depends_on('python@:3.5,3.7:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-cgen/package.py]: Where do you see that Python 3.6 doesn't work?

File: var/spack/repos/builtin/packages/py-codepy/package.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    pypi     = "codepy/codepy-2019.1.tar.gz"
+
+    version('2019.1', sha256='384f22c37fe987c0ca71951690c3c2fd14dacdeddbeb0fde4fd01cd84859c94e')
+
+    depends_on('python@:3.5,3.7:', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
+    depends_on('py-pytools@2015.1.2:', type='build')
+    depends_on('py-numpy@1.6:', type='build')
+    depends_on('py-platformdirs@2.2.0:', type='build')
+    depends_on('py-cgen', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-codepy/package.py]: Same comments here

File: var/spack/repos/builtin/packages/py-devito/package.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    pypi     = "devito/devito-4.6.2.tar.gz"
+
+    version('4.6.2', sha256='39c2210a192ad69953b4f8d93440ffd72b07d739c4fe2290e2b182adfb7e143f')
+
+    variant('mpi', default=False, description='Enable MPI support')
+    variant('matplotlib', default=False, description='Enable matplolib support')
+    variant('pandas', default=False, description='Enable pandas support')
+
+    depends_on('py-pip@9.0.1:', type='build')
+    depends_on('py-versioneer', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-devito/package.py]: Where do you see a versioneer dep?

File: var/spack/repos/builtin/packages/py-devito/package.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    depends_on('py-setuptools', type='build')
+
+    depends_on('py-numpy@1.16:', type=('build', 'run'))
+    depends_on('py-sympy@1.7:1.9', type=('build', 'run'))
+    depends_on('py-scipy', type=('build', 'run'))
+    depends_on('py-flake8@2.1.0:', type=('build', 'run'))
+    depends_on('py-nbval', type=('build', 'run'))
+    depends_on('py-cached-property', type=('build', 'run'))
+    depends_on('py-psutil@5.1.0:5.99', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-devito/package.py]:
```suggestion
    depends_on('py-psutil@5.1.0:5', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-devito/package.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    depends_on('py-py-cpuinfo@8:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-devito/package.py]:
```suggestion
    depends_on('py-py-cpuinfo@:8', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-devito/package.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    depends_on('py-cgen@2020.1:', type=('build', 'run'))
+    depends_on('py-codepy@2019.1:', type=('build', 'run'))
+    depends_on('py-click@:8.99', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-devito/package.py]:
```suggestion
    depends_on('py-click@:8', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-devito/package.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    depends_on('py-codecov', type=('build', 'run'))
+    depends_on('py-multidict', type=('build', 'run'))
+    depends_on('py-anytree@2.4.3:2.8', type=('build', 'run'))
+    depends_on('py-pyrevolve@2.1.3:', type=('build', 'run'))
+    depends_on('py-distributed@2022.2:', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-devito/package.py]:
```suggestion
    depends_on('py-distributed@:2022.2', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-devito/package.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    depends_on('py-pytest@3.6:7.99', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-devito/package.py]:
```suggestion
    depends_on('py-pytest@3.6:6', type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-devito/package.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    depends_on('py-pytest-runner', type=('build', 'run'))
+    depends_on('py-pytest-cov', type=('build', 'run'))
+
+    # requirements-mpi.tct
+    depends_on('py-mpi4py@:4.0', type=('build', 'run'), when='+mpi')
+    depends_on('py-ipyparallel@:8.4', type=('build', 'run'), when='+mpi')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-devito/package.py]: These version constraints are only on master, right? I don't think they are constrained in the latest stable release.

File: var/spack/repos/builtin/packages/py-devito/package.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+    # requirements-optional.txt
+    depends_on('py-matplotlib', type=('build', 'run'), when='+matplotlib')
+    depends_on('py-pandas', type=('build', 'run'), when='+pandas')
+
+    depends_on('mpi', type=('build', 'run'))
+
+    depends_on('intel-parallel-studio', type='run', when='%intel@:2021.1.1')
+    depends_on('intel-oneapi-compilers', type='run', when='%intel@2021.1.2:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-devito/package.py:62]: Compilers aren't currently modeled as dependencies. You can build the Intel compilers in Spack, but you don't add them as dependencies to packages (yet). I would remove these lines.

File: var/spack/repos/builtin/packages/py-nbval/package.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyNbval(PythonPackage):
+    """Py.test plugin for validating Jupyter notebooks.
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbval/package.py:10]: Oh cool, didn't know about this one! I've been using nbmake for testing.

File: var/spack/repos/builtin/packages/py-nbval/package.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    inputs match the stored outputs of the .ipynb file. Whilst also ensuring that the
+    notebooks are running without errors.
+    """
+
+    homepage = "https://github.com/computationalmodelling/nbval"
+    pypi     = "nbval/nbval-0.9.6.tar.gz"
+
+    version('0.9.6', sha256='cfefcd2ef66ee2d337d0b252c6bcec4023384eb32e8b9e5fcc3ac80ab8cd7d40')
+
+    depends_on('python@3.6:3.99', type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbval/package.py]: Seems like they still support 2, I would remove this line

File: var/spack/repos/builtin/packages/py-nbval/package.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    depends_on('py-setuptools', type='build')
+    depends_on('py-pytest@2.8:', type=('build'))
+    depends_on('py-jupyter-client', type=('build'))
+    depends_on('py-nbformat', type=('build'))
+    depends_on('py-ipykernel', type=('build'))
+    depends_on('py-coverage', type=('build'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-nbval/package.py]: All deps should be type=build/run, missing six dep for Python 2 support

File: var/spack/repos/builtin/packages/py-pyrevolve/package.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+    homepage = "https://github.com/devitocodes/pyrevolve"
+    pypi     = "pyrevolve/pyrevolve-2.2.tar.gz"
+
+    version('2.2', sha256='b49aea5cd6c520ac5fcd1d25fa23fe2c5502741d2965f3eee10be067e7b0efb4')
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-contexttimer', type='build')
+    depends_on('py-cython@0.17:', type='build')
+    depends_on('py-flake8', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyrevolve/package.py]: Flake8 not required, but versioneer is

File: var/spack/repos/builtin/packages/py-pyrevolve/package.py [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+    Python wrapper for Revolve: https://dl.acm.org/doi/10.1145/347837.347846
+    """
+
+    homepage = "https://github.com/devitocodes/pyrevolve"
+    pypi     = "pyrevolve/pyrevolve-2.2.tar.gz"
+
+    version('2.2', sha256='b49aea5cd6c520ac5fcd1d25fa23fe2c5502741d2965f3eee10be067e7b0efb4')
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-contexttimer', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-pyrevolve/package.py]: Should be type=build/run

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-cgen/package.py (hunk 2 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-cgen/package.py]: This was in the main brnach, removing.

File: var/spack/repos/builtin/packages/py-cgen/package.py (hunk 1 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-cgen/package.py]: Done

File: var/spack/repos/builtin/packages/py-codepy/package.py (hunk 3 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-codepy/package.py]: Changed it. Again, I wronlgy used the build dependencies from the main branch.

File: var/spack/repos/builtin/packages/py-devito/package.py (hunk 4 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-devito/package.py]: Here: https://github.com/devitocodes/devito/blob/144c4bb09facc7402e2684c9d2d230c5a530dee8/setup.py#L1

Could this also be resolved by the local `versioneer.py` inside the project?

File: var/spack/repos/builtin/packages/py-devito/package.py (hunk 5 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-devito/package.py]: Done.

File: var/spack/repos/builtin/packages/py-nbval/package.py (hunk 13 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-nbval/package.py]: Again, I got that from the master brnach. There seems to be some consistency to my errors.

File: var/spack/repos/builtin/packages/py-nbval/package.py (hunk 14 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-nbval/package.py]: done.

File: var/spack/repos/builtin/packages/py-pyrevolve/package.py (hunk 15 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-pyrevolve/package.py]: Done.  And made contexttimer dep `('build', 'run')` as is used in `install_requires`

File: var/spack/repos/builtin/packages/py-devito/package.py (hunk 6 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-devito/package.py]: Done

File: var/spack/repos/builtin/packages/py-devito/package.py (hunk 7 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-devito/package.py]: done

File: var/spack/repos/builtin/packages/py-devito/package.py (hunk 8 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-devito/package.py]: Done

File: var/spack/repos/builtin/packages/py-devito/package.py (hunk 9 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-devito/package.py]: Done.

File: var/spack/repos/builtin/packages/py-devito/package.py (hunk 10 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-devito/package.py]: True. Again, I used the deps from master. Changed them.

File: var/spack/repos/builtin/packages/py-devito/package.py (hunk 11 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-devito/package.py:62]: I actually need `icc` during run time. `benchmarks/user/benchmark.py` will compile stuff on the fly while running.

File: var/spack/repos/builtin/packages/py-devito/package.py (hunk 4 above)
Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-devito/package.py]: Seems to build fine with internal `versioneer.py`. I'll remove external dependency.

**Review by adamjstewart — DISMISSED** [review]

File: var/spack/repos/builtin/packages/py-codepy/package.py [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    - Compiling this source code and dynamically loading it into the Python interpreter.
+    """
+
+    homepage = "https://documen.tician.de/codepy/"
+    pypi     = "codepy/codepy-2019.1.tar.gz"
+
+    version('2019.1', sha256='384f22c37fe987c0ca71951690c3c2fd14dacdeddbeb0fde4fd01cd84859c94e')
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-pytools@2015.1.2:', type=('build','run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-codepy/package.py]:
```suggestion
    depends_on('py-pytools@2015.1.2:', type=('build', 'run'))
```
Same for other deps and in py-nbval, this is causing the style tests to fail

Comment (stephenmsachs) [reply var/spack/repos/builtin/packages/py-codepy/package.py]: Done.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**